### DISCUSSION TITLE
post/best-python-dev-tooling/

### DISCUSSION THREAD

**Discussion by giscus** [basnijholt/nijho.lt#14] [General]
# post/best-python-dev-tooling/

Explore the top developer tools for enhancing productivity and code quality in Python projects

https://www.nijho.lt/post/best-python-dev-tooling/

**Comment by basnijholt** [comment]
Let me know what you think!

**Comment by aeantipov** [comment]
Bas this is a nice blog :) 
I always felt that mypy is an attempt to fix the weak typing of python. I.e. it just puts a standard to have an explicitly typed code, however without a compiler to use this

**Reply by basnijholt** [reply]
The loose typing is also quite convenient in many cases, however, it leads to hard to understand code sometimes.

One thing I learned recently, I will never develop new Python code without using a static type checker! It just identifier so many problems before even running the code 😄

**Comment by cityhunteur** [comment]
Great article. Wish I came across it earlier :) I'm refreshing my Python atm and ended up with the same set of tools/workflow after couple of days. Trying to get as close to the Go experience (fast feedback, typing, etc.) 

Also came across [pyink](https://github.com/google/pyink); a fork of `black` which adheres to Google style guides.

**Comment by aeantipov** [comment]
@basnijholt, I noticed you updated your guide to include uv, great! For this one - I found that versioningit is a bit of an overkill, since you now have `hatch` and [`hatch-vcs`](https://github.com/ofek/hatch-vcs), which generates a `_version.py` on pip install instead of you having to write a proto.

**Reply by basnijholt** [reply]
Thanks for coming back! 😄 

The reason I still use `versioningit` is because it provides an updated version at runtime, so `package.__version__` can still include the latest commit hash, which is useful when working with `--editable` installs. With `hatch-vcs` it keeps the version as when it was installed.
